ipc/provider/src/lib.rs

```diff
@@ -263,63 +263,144 @@ impl IpcProvider {
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        let worker = match worker_addr {
+            None => sender,
+            Some(addr) => addr,
+        };
+        conn.manager()
+            .join_subnet(subnet, sender, collateral, validator_net_addr, worker)
+            .await
     }
 
-    /// Sends a request to leave a subnet from a wallet address.
-    pub async fn leave_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn leave_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().leave_subnet(subnet, sender).await
     }
 
-    /// Sends a signal to kill a subnet
-    pub async fn kill_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn kill_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().kill_subnet(subnet, sender).await
     }
 
-    /// Lists all the registered children in a gateway.
     pub async fn list_child_subnets(
         &self,
-        _gateway_addr: Address,
+        gateway_addr: Option<Address>,
+        subnet: &SubnetID,
     ) -> anyhow::Result<HashMap<SubnetID, SubnetInfo>> {
-        todo!()
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+
+        let gateway_addr = match gateway_addr {
+            None => subnet_config.gateway_addr(),
+            Some(addr) => addr,
+        };
+
+        conn.manager().list_child_subnets(gateway_addr).await
```

Comment:
Gateway address should not be needed in `list_child_subnets` as `manager` should already know which subnet it is `managing` and hence should have already known the `gateway` address. Ideally, manager should have owned the `gateway` address?.

Reply:
As this is a huge change, I will fix it in a follow up PR. Tracked in this issue: https://github.com/consensus-shipyard/ipc-agent/issues/316